Compare only fields of given type

beyond-danube opened the pull request:
#### Check List:
* Fixes #2793
* Unit tests : YES
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj/blob/main/CONTRIBUTING.md) YES

beyond-danube commented:
#### Short description
Introduce new field in Recursive Comparison Configuration - set of types to be compared, and include it to existing places (description, equals, etc.)

Keep track of parent fields in separate field. Since when matching `DualValue` we don't know if parent type should be compared, so we do that based on matching set of parent paths. Maybe not a best place to track this, however it does not have accessors so cannot be messed up unintentionally.

We should evaluate existing `comparedFields` together with the new `comparedFieldsOfTypes` (or any other further matching based on value) since fields might filter out each other otherwise.

This change does not break any of existing tests so it's safe I belive.

joel-costigliola commented:
Thanks, I will review your PR in the next few days

joel-costigliola commented on assertj-core/src/test/java/org/assertj/core/api/recursive/comparison/RecursiveComparisonAssert_isEqualTo_comparingOnlyFieldsOfTypes_Test.java:
```diff
@@ -0,0 +1,221 @@
+
+    // THEN
+    verifyShouldBeEqualByComparingFieldByFieldRecursivelyCall(actual, expected,
+                                                              differences);
```
minor: join lines

joel-costigliola reviewed (commented):
Thanks @beyond-danube nice PR, first round of code review done.

beyond-danube commented:
@joel-costigliola thanks for the review! I will go through your points in the nearest few days

beyond-danube commented on assertj-core/src/main/java/org/assertj/core/api/recursive/comparison/RecursiveComparisonConfiguration.java:
```diff
@@ -792,6 +835,19 @@ private boolean matchesAnIgnoredFieldType(DualValue dualValue) {
   }
 
+  private boolean shouldBeComparedBasedOnFieldValue(DualValue dualValue) {
+    if (parentDualValuePathsComparedFieldsOfTypes.stream().anyMatch(path -> dualValue.fieldLocation.getPathToUseInRules().startsWith(path))) return matchFiledComparedByTypeAndStorePath(dualValue);
```
Totally agree here, seems that I have just missed this capability in my initial implementation. It's more obvious way to store filed locations indeed so I have done that.

beyond-danube commented on assertj-core/src/main/java/org/assertj/core/api/recursive/comparison/RecursiveComparisonConfiguration.java:
```diff
@@ -792,6 +835,19 @@ private boolean matchesAnIgnoredFieldType(DualValue dualValue) {
+  private boolean shouldBeComparedBasedOnFieldValue(DualValue dualValue) {
+    if (parentDualValuePathsComparedFieldsOfTypes.stream().anyMatch(path -> dualValue.fieldLocation.getPathToUseInRules().startsWith(path))) return matchFiledComparedByTypeAndStorePath(dualValue);
+    if (dualValue.actual != null && comparedFieldsOfTypes.contains(dualValue.actual.getClass())) return matchFiledComparedByTypeAndStorePath(dualValue);
+    if (dualValue.expected != null && comparedFieldsOfTypes.contains(dualValue.expected.getClass())) return matchFiledComparedByTypeAndStorePath(dualValue);
```
Added javadoc on how is compared type evaluated.

We don't evaluate `dualValue.expected` always now, we'll get to the `dualValue.expected`'s `if` statement only if `dualValue.actual` is `null`

However we can remove `matchFiledComparedByTypeAndStorePath(DualValue dualValue)` method, which only adds `FieldLocation` to parent compared fields set and always returns `true` and modify `shouldBeComparedBasedOnFieldValue(DualValue dualValue)` to something like:

```java
  private boolean shouldBeComparedBasedOnFieldValue(DualValue dualValue) {
    if (fieldLocationOfComparedFieldsOfTypes.stream().anyMatch(dualValue.fieldLocation::hasParent)) {
      fieldLocationOfComparedFieldsOfTypes.add(dualValue.fieldLocation);
      return true;
    }
    if (dualValue.actual != null && comparedFieldsOfTypes.contains(dualValue.actual.getClass())) {
      fieldLocationOfComparedFieldsOfTypes.add(dualValue.fieldLocation);
      return true;
    }
    if (dualValue.expected != null && comparedFieldsOfTypes.contains(dualValue.expected.getClass())) {
      fieldLocationOfComparedFieldsOfTypes.add(dualValue.fieldLocation);
      return true;
    }

    return false;
  }
```

So that might be more readable.

To be honest I don't have any strong argument for any approach here so we can just pick one - more obvious or less repetitiveness.

beyond-danube commented:
Hi @joel-costigliola I did go though your points and added changes on all of them, also spotted a typo and fixed that as well. Please have a look at the changes and [this part](https://github.com/assertj/assertj/pull/2794#discussion_r992807129).

joel-costigliola commented:
Integrated thanks @beyond-danube!

beyond-danube commented:
Great, thanks @joel-costigliola!
